### PR TITLE
PR1 of conditional BP

### PR DESCRIPTION
**Pull request by andxu**
1. Basic flow to handle breakpoint and condition.

### CODE REVIEW

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java [hunk 1]
```diff
@@ -73,6 +84,11 @@ public void initialize(IDebugAdapterContext context) {
             return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.EMPTY_DEBUG_SESSION, "Empty debug session.");
         }
 
+        if (!registered) {
```
Comment (yaohaizh) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: Should synchronize the register behavior.

Comment (yaohaizh) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: Should register behavior happen on the initialization?

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java [hunk 2]
```diff
@@ -130,6 +146,55 @@ public void initialize(IDebugAdapterContext context) {
...
+                }
+                Event event = debugEvent.event;
+                if (debugEvent.eventSet.size() > 1 && debugEvent.eventSet.stream().anyMatch(t -> t instanceof StepEvent)) {
+                    // The StepEvent and BreakpointEvent are grouped in the same event set only if they occurs at the same location and in the same thread.
+                    // In order to avoid two duplicated StoppedEvents, the debugger will skip the BreakpointEvent.
+                } else {
+                    ThreadReference bpThread = ((BreakpointEvent) event).thread();
+                    IEvaluationProvider engine = context.getProvider(IEvaluationProvider.class);
+                    if (engine.isInEvaluation(bpThread)) {
+                        return;
```
Comment (yaohaizh) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: If early return, what is the expected value of debugEvent.shouldResume?

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java (hunk 1 above)
Comment (andxu) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: on initialization, the debug session is not ready, always null in initialize method

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java (hunk 2 above)
Comment (andxu) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: shouldResume is default true, that is expected.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java (hunk 1 above)
Comment (akaroml) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: There is a initialize member of IDebugRequestHandler to handle the initialization. This logic should be there.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java [hunk 3]
```diff
@@ -130,6 +146,55 @@ public void initialize(IDebugAdapterContext context) {
         }
     }
 
+    private void registerBreakpointHandler(IDebugAdapterContext context) {
+        IDebugSession debugSession = context.getDebugSession();
+        if (debugSession != null) {
+            debugSession.getEventHub().events().subscribe(debugEvent -> {
+                if (!(debugEvent.event instanceof BreakpointEvent)) {
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: use observable.filter to only process BreakpointEvent.

Comment (andxu) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java]: Done in PR2

**Pull request merged**